lib/db/schemaupdater.go

```diff
@@ -421,24 +429,69 @@ func (db *schemaUpdater) updateSchema6to7(_ int) error {
 	return t.Commit()
 }
 
-func (db *schemaUpdater) updateSchema7to8(_ int) error {
+func (db *schemaUpdater) updateSchemato9(prev int) error {
 	// Loads and rewrites all files with blocks, to deduplicate block lists.
+	// Checks for missing or incorrect sequence entries and rewrites those.
 
 	t, err := db.newReadWriteTransaction()
 	if err != nil {
 		return err
 	}
 	defer t.close()
 
+	var sk []byte
 	it, err := t.NewPrefixIterator([]byte{KeyTypeDevice})
 	if err != nil {
 		return err
 	}
+	metas := make(map[string]*metadataTracker)
 	for it.Next() {
 		var fi protocol.FileInfo
 		if err := fi.Unmarshal(it.Value()); err != nil {
 			return err
 		}
+		device, ok := t.keyer.DeviceFromDeviceFileKey(it.Key())
+		if !ok {
+			return errDeviceIdxMissing
+		}
+		if bytes.Equal(device, protocol.LocalDeviceID[:]) {
+			folder, ok := t.keyer.FolderFromDeviceFileKey(it.Key())
+			if !ok {
+				return errFolderIdxMissing
+			}
+			if sk, err = t.keyer.GenerateSequenceKey(sk, folder, fi.Sequence); err != nil {
+				return err
+			}
+			switch dk, err := t.Get(sk); {
+			case err != nil:
+				if !backend.IsNotFound(err) {
+					return err
+				}
+				fallthrough
+			case !bytes.Equal(it.Key(), dk):
+				folderStr := string(folder)
+				meta, ok := metas[folderStr]
+				if !ok {
+					meta = loadMetadataTracker(db.Lowlevel, folderStr)
+					metas[folderStr] = meta
+				}
+				fi.Sequence = meta.nextLocalSeq()
+				if sk, err = t.keyer.GenerateSequenceKey(sk, folder, fi.Sequence); err != nil {
+					return err
+				}
+				if err := t.Put(sk, it.Key()); err != nil {
+					return err
+				}
+				if err := t.putFile(it.Key(), fi); err != nil {
```

Comment:
Actually this thing is broken. We unmarshal to a raw fileinfo, which will have a block hash but not filled in blocks, then use putFile which expects a fully formed FileInfo with blocks filled in. The combo is bad and will result in brokenness. Either we need to actually load the blocks as well using unmarshalTrunc or do a raw marshal+put as well.

Reply:
At least now when putFile reacts badly to `blocks == nil && blocksHash != nil`. If we want it to be able to save raw FileInfos then it must let that past without complains, including not trying to nil out blocksHash just because the blocks are nil.

Reply:
Oh crap, nice find. Makes the panics almost benign compared to wrecking dbs with the latest, luckily unshipped change of niling blockshashes.

I agree ignoring blockhash consistency at such a low level makes most sense.

And I need to stop any future urge to use a panic to ensure expected behavior. Either it's important enough to test or add something like "non-panic sentry events" or just behave sanely in any case...